File: src/commands/entity_timeline.rs
//! Entity timeline visualization command.
//!
//! ⚠️ **DEPRECATED**: Use `entity-analysis timeline` instead.
//!
//! ```bash
//! # Old (deprecated):
//! vault-audit entity-timeline logs/*.log --entity-id abc-123-def
//!
//! # New (recommended):
//! vault-audit entity-analysis timeline --entity-id abc-123-def logs/*.log
//! ```
//!
//! See [`entity_analysis`](crate::commands::entity_analysis) for the unified command.
//!
//! ---
//!
//! Generates a detailed timeline of all operations performed by a specific entity,
//! useful for understanding entity behavior and troubleshooting issues.
//! Supports multi-file analysis to track entities across multiple days.
//!
//! # Usage
//!
//! ```bash
//! # Single file
//! vault-audit entity-timeline audit.log --entity-id abc-123-def
//!
//! # Multi-day timeline
//! vault-audit entity-timeline day1.log day2.log day3.log --entity-id abc-123-def
//! ```
//!
//! # Output
//!
//! Displays a chronological view of the entity's activity:
//! - Timestamp
//! - Operation type (read, write, list, etc.)
//! - Path accessed
//! - Response status
//!
//! Also provides:
//! - Activity summary (operations by type)
//! - Time-based patterns (hourly distribution)
//! - Mount point usage
//! - First and last seen timestamps

use crate::audit::types::AuditEntry;
use crate::utils::format::format_number;
use crate::utils::progress::ProgressBar;
use crate::utils::reader::open_file;
use anyhow::Result;
use chrono::{DateTime, Timelike, Utc};
use std::collections::HashMap;
use std::io::{BufRead, BufReader};

#[derive(Clone)]
#[allow(dead_code)]
struct Operation {
    timestamp: DateTime<Utc>,
    path: String,
    op: String,
}

pub fn run(log_files: &[String], entity_id: &str, display_name: Option<&String>) -> Result<()> {
    println!("Analyzing timeline for entity: {}", entity_id);
    if let Some(name) = display_name {
        println!("Display name: {}", name);
    }
    println!();

    let mut operations_by_hour: HashMap<String, HashMap<String, usize>> = HashMap::new();
    let mut operations_by_type: HashMap<String, usize> = HashMap::new();
    let mut paths_accessed: HashMap<String, usize> = HashMap::new();
    let mut operations_timeline: Vec<Operation> = Vec::new();
    let mut total_lines = 0;
    let mut entity_operations = 0;

    // Process each log file sequentially
    for (file_idx, log_file) in log_files.iter().enumerate() {
        eprintln!(
            "[{}/{}] Processing: {}",
            file_idx + 1,
            log_files.len(),
            log_file
        );

        // Count lines in file first for accurate progress tracking
        eprintln!("Scanning file to determine total lines...");
        let total_file_lines = crate::utils::parallel::count_file_lines(log_file)?;

        let progress = ProgressBar::new(total_file_lines, "Processing");

        let file = open_file(log_file)?;
        let reader = BufReader::new(file);

        let mut file_lines = 0;

        for line in reader.lines() {
            file_lines += 1;
            total_lines += 1;
            let line = line?;

            if file_lines % 10_000 == 0 {
                progress.update(file_lines);
            }

            let entry: AuditEntry = match serde_json::from_str(&line) {
                Ok(e) => e,
                Err(_) => continue,
            };

            // Check if this is our entity
            let entry_entity_id = match &entry.auth {
                Some(auth) => match &auth.entity_id {
                    Some(id) => id.as_str(),
                    None => continue,
                },
                None => continue,
            };

            if entry_entity_id != entity_id {
                continue;
            }

            entity_operations += 1;

            let path = entry
                .request
                .as_ref()
                .and_then(|r| r.path.as_deref())
                .unwrap_or("")
                .to_string();
            let operation = entry
                .request
                .as_ref()
                .and_then(|r| r.operation.as_deref())
                .unwrap_or("")
                .to_string();

            if let Ok(ts) = chrono::DateTime::parse_from_rfc3339(&entry.time) {
                let ts_utc = ts.with_timezone(&Utc);

                // Track by hour
                let hour_key = ts_utc.format("%Y-%m-%d %H:00").to_string();
                let hour_ops = operations_by_hour.entry(hour_key).or_default();
                *hour_ops.entry("total".to_string()).or_insert(0) += 1;
                *hour_ops.entry(operation.clone()).or_insert(0) += 1;

                // Store operation for timeline
                operations_timeline.push(Operation {
                    timestamp: ts_utc,
                    path: path.clone(),
                    op: operation.clone(),
                });
            }

            // Track operation types
            *operations_by_type.entry(operation).or_insert(0) += 1;

            // Track paths
            *paths_accessed.entry(path).or_insert(0) += 1;
        }

        // Ensure 100% progress for this file
        progress.update(total_file_lines);

        progress.finish_with_message(&format!(
            "Processed {} lines from this file",
            format_number(file_lines)
        ));
    }

    eprintln!(
        "\nTotal: Processed {} lines, found {} operations for entity: {}",
        format_number(total_lines),
        format_number(entity_operations),
        entity_id
    );

    if entity_operations == 0 {
        println!("\nNo operations found for this entity!");
        return Ok(());
    }

    // Sort timeline
    operations_timeline.sort_by(|a, b| a.timestamp.cmp(&b.timestamp));

    // Calculate time span
    let (first_op, last_op, time_span_hours) = if operations_timeline.is_empty() {
        return Ok(());
    } else {
        let first = operations_timeline.first().unwrap().timestamp;
        let last = operations_timeline.last().unwrap().timestamp;
        let span = (last - first).num_seconds() as f64 / 3600.0;
        (first, last, span)
    };

    // Analysis and reporting
    println!("\n{}", "=".repeat(100));
    println!("TIMELINE ANALYSIS FOR: {}", entity_id);
    println!("{}", "=".repeat(100));

    // 1. Summary statistics
    println!("\n1. SUMMARY STATISTICS");
    println!("{}", "-".repeat(100));
    println!("Total operations: {}", format_number(entity_operations));
    println!(
        "Time span: {:.2} hours ({:.2} days)",
        time_span_hours,
        time_span_hours / 24.0
    );
    println!(
        "Average rate: {:.1} operations/hour ({:.2}/minute)",
        entity_operations as f64 / time_span_hours,
        entity_operations as f64 / time_span_hours / 60.0
    );
    println!("First operation: {}", first_op.format("%Y-%m-%d %H:%M:%S"));
    println!("Last operation: {}", last_op.format("%Y-%m-%d %H:%M:%S"));

    // 2. Operation type distribution
    println!("\n2. OPERATION TYPE DISTRIBUTION");
    println!("{}", "-".repeat(100));
    println!("{:<30} {:<15} {:<15}", "Operation", "Count", "Percentage");
    println!("{}", "-".repeat(100));

    let mut sorted_ops: Vec<_> = operations_by_type.iter().collect();
    sorted_ops.sort_by(|a, b| b.1.cmp(a.1));

    for (op, count) in sorted_ops {
        let percentage = (*count as f64 / entity_operations as f64) * 100.0;
        println!(
            "{:<30} {:<15} {:<15.2}%",
            op,
            format_number(*count),
            percentage
        );
    }

    // 3. Top paths accessed
    println!("\n3. TOP 30 PATHS ACCESSED");
    println!("{}", "-".repeat(100));
    println!("{:<70} {:<15} {:<15}", "Path", "Count", "Percentage");
    println!("{}", "-".repeat(100));

    let mut sorted_paths: Vec<_> = paths_accessed.iter().collect();
    sorted_paths.sort_by(|a, b| b.1.cmp(a.1));

    for (path, count) in sorted_paths.iter().take(30) {
        let percentage = (**count as f64 / entity_operations as f64) * 100.0;
        let display_path = if path.len() > 68 {
            format!("{}...", &path[..65])
        } else {
            (*path).clone()
        };
        println!(
            "{:<70} {:<15} {:<15.2}%",
            display_path,
            format_number(**count),
            percentage
        );
    }

    // 4. Hourly activity pattern
    println!("\n4. HOURLY ACTIVITY PATTERN (Top 30 Hours)");
    println!("{}", "-".repeat(100));
    println!(
        "{:<20} {:<12} {:<10} {:<10} {:<10} {:<10}",
        "Hour", "Total Ops", "read", "update", "list", "Other"
    );
    println!("{}", "-".repeat(100));

    let mut sorted_hours: Vec<_> = operations_by_hour.iter().collect();
    sorted_hours.sort_by(|a, b| {
        let a_total = a.1.get("total").unwrap_or(&0);
        let b_total = b.1.get("total").unwrap_or(&0);
        b_total.cmp(a_total)
    });

    for (hour, ops) in sorted_hours.iter().take(30) {
        let total = *ops.get("total").unwrap_or(&0);
        let read = *ops.get("read").unwrap_or(&0);
        let update = *ops.get("update").unwrap_or(&0);
        let list_operations = *ops.get("list").unwrap_or(&0);
        let other = total - read - update - list_operations;

        println!(
            "{:<20} {:<12} {:<10} {:<10} {:<10} {:<10}",
            hour,
            format_number(total),
            format_number(read),
            format_number(update),
            format_number(list_operations),
            format_number(other)
        );
    }

    // 5. Activity distribution by hour of day
    println!("\n5. ACTIVITY DISTRIBUTION BY HOUR OF DAY");
    println!("{}", "-".repeat(100));

    let mut hour_of_day_stats: HashMap<u32, usize> = HashMap::new();
    for op in &operations_timeline {
        let hour = op.timestamp.hour();
        *hour_of_day_stats.entry(hour).or_insert(0) += 1;
    }

    println!("{:<10} {:<15} {:<50}", "Hour", "Operations", "Bar Chart");
    println!("{}", "-".repeat(100));

    let max_ops_in_hour = hour_of_day_stats.values().max().copied().unwrap_or(1);

    for hour in 0..24 {
        let ops = *hour_of_day_stats.get(&hour).unwrap_or(&0);
        let bar_length = if max_ops_in_hour > 0 {
            (ops * 50) / max_ops_in_hour
        } else {
            0
        };
        let bar = "█".repeat(bar_length);
        println!("{:02}:00     {:<15} {}", hour, format_number(ops), bar);
    }

    // 6. Peak activity analysis
    println!("\n6. PEAK ACTIVITY WINDOWS");
    println!("{}", "-".repeat(100));

    let mut window_counts: HashMap<DateTime<Utc>, usize> = HashMap::new();

    for op in &operations_timeline {
        // Round to 5-minute window
        let minute = (op.timestamp.minute() / 5) * 5;
        let window_start = op
            .timestamp
            .with_minute(minute)
            .unwrap()
            .with_second(0)
            .unwrap()
            .with_nanosecond(0)
            .unwrap();
        *window_counts.entry(window_start).or_insert(0) += 1;
    }

    let mut sorted_windows: Vec<_> = window_counts.iter().collect();
    sorted_windows.sort_by(|a, b| b.1.cmp(a.1));

    println!(
        "{:<25} {:<15} {:<20}",
        "5-Minute Window", "Operations", "Rate (ops/sec)"
    );
    println!("{}", "-".repeat(100));

    for (window, count) in sorted_windows.iter().take(20) {
        let rate = **count as f64 / 300.0;
        println!(
            "{:<25} {:<15} {:<20.3}",
            window.format("%Y-%m-%d %H:%M"),
            format_number(**count),
            rate
        );
    }

    // 7. Behavioral patterns
    println!("\n7. BEHAVIORAL PATTERNS");
    println!("{}", "-".repeat(100));

    if time_span_hours > 1.0 {
        let ops_per_hour = entity_operations as f64 / time_span_hours;
        if ops_per_hour > 100.0 {
            println!(
                "⚠️  HIGH FREQUENCY: {:.0} operations/hour suggests automated polling",
                ops_per_hour
            );
            println!("   Recommended action: Implement caching or increase polling interval");
        }

        // Check for token lookup abuse
        let token_lookup_paths: Vec<_> = paths_accessed
            .keys()
            .filter(|p| p.contains("token/lookup"))
            .collect();
        let total_token_lookups: usize = token_lookup_paths
            .iter()
            .map(|p| paths_accessed.get(*p).unwrap_or(&0))
            .sum();

        if total_token_lookups > 1000 {
            println!(
                "⚠️  TOKEN LOOKUP ABUSE: {} token lookups detected",
                format_number(total_token_lookups)
            );
            println!(
                "   Rate: {:.1} lookups/hour = {:.2} lookups/second",
                total_token_lookups as f64 / time_span_hours,
                total_token_lookups as f64 / time_span_hours / 3600.0
            );
            println!("   Recommended action: Implement client-side token TTL tracking");
        }

        // Check for path concentration
        if let Some((top_path, top_count)) = sorted_paths.first() {
            let top_path_pct = (**top_count as f64 / entity_operations as f64) * 100.0;
            if top_path_pct > 30.0 {
                println!(
                    "⚠️  PATH CONCENTRATION: {:.1}% of operations on single path",
                    top_path_pct
                );
                println!("   Path: {}", top_path);
                println!(
                    "   Recommended action: Review why this path is accessed {} times",
                    format_number(**top_count)
                );
            }
        }

        // Check for 24/7 activity
        let hours_with_activity = (0..24)
            .filter(|h| hour_of_day_stats.contains_key(h))
            .count();
        if hours_with_activity >= 20 {
            println!(
                "⚠️  24/7 ACTIVITY: Active in {}/24 hours",
                hours_with_activity
            );
            println!("   Suggests automated system or background process");
        }
    }

    println!("\n{}", "=".repeat(100));

    Ok(())
}

File: internal/commands/entity_timeline.go
// Package commands provides subcommands for vault audit analysis.
package commands

import (
	"fmt"
	"os"
	"sort"
	"strings"
	"time"

	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/audit"
	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/processor"
	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/utils"
)

// Operation represents a single operation in an entity's timeline.
type Operation struct {
	timestamp time.Time
	path      string
	op        string
}

// TimelineState tracks operations for a specific entity.
type TimelineState struct {
	operations      []Operation
	operationsByType map[string]int
	pathsAccessed    map[string]int
	operationsByHour map[string]map[string]int
	hourOfDayStats   map[int]int
	windowCounts     map[time.Time]int
	entityOpCount    int
}

// RunEntityTimeline shows chronological activity for a specific entity.
func RunEntityTimeline(logFiles []string, entityID string, displayName *string) error {
	fmt.Printf("Analyzing timeline for entity: %s\n", entityID)
	if displayName != nil && *displayName != "" {
		fmt.Printf("Display name: %s\n", *displayName)
	}
	fmt.Printf("\n")

	result, _, err := processor.RunFiles(
		processor.DefaultConfig(),
		logFiles,
		func() TimelineState {
			return TimelineState{
				operations:       make([]Operation, 0),
				operationsByType: make(map[string]int),
				pathsAccessed:    make(map[string]int),
				operationsByHour: make(map[string]map[string]int),
				hourOfDayStats:   make(map[int]int),
				windowCounts:     make(map[time.Time]int),
			}
		},
		func(entry *audit.AuditEntry, s *TimelineState) {
			// Check if this is our entity
			if entry.EntityID() != entityID {
				return
			}

			s.entityOpCount++

			path := entry.Path()
			op := entry.Operation()

			ts := parseTime(entry.Time)

			// Track by hour
			hourKey := ts.Format("2006-01-02 15:00")
			if _, ok := s.operationsByHour[hourKey]; !ok {
				s.operationsByHour[hourKey] = make(map[string]int)
			}
			s.operationsByHour[hourKey]["total"]++
			s.operationsByHour[hourKey][op]++

			// Store operation for timeline
			s.operations = append(s.operations, Operation{
				timestamp: ts,
				path:      path,
				op:        op,
			})

			// Track operation types
			s.operationsByType[op]++

			// Track paths
			s.pathsAccessed[path]++

			// Track by hour of day
			s.hourOfDayStats[int(ts.Hour())]++

			// Track 5-minute windows
			minute := (ts.Minute() / 5) * 5
			windowStart := ts.Truncate(time.Hour)
			windowStart = windowStart.Add(time.Duration(minute) * time.Minute)
			s.windowCounts[windowStart]++
		},
		func(a, b TimelineState) TimelineState {
			// Merge states
			a.operations = append(a.operations, b.operations...)
			for op, count := range b.operationsByType {
				a.operationsByType[op] += count
			}
			for path, count := range b.pathsAccessed {
				a.pathsAccessed[path] += count
			}
			for hour, ops := range b.operationsByHour {
				if _, ok := a.operationsByHour[hour]; !ok {
					a.operationsByHour[hour] = make(map[string]int)
				}
				for op, count := range ops {
					a.operationsByHour[hour][op] += count
				}
			}
			for hour, count := range b.hourOfDayStats {
				a.hourOfDayStats[hour] += count
			}
			for window, count := range b.windowCounts {
				a.windowCounts[window] += count
			}
			a.entityOpCount += b.entityOpCount
			return a
		},
	)

	if err != nil {
		return err
	}

	entityOps := result.entityOpCount

	fmt.Fprintf(os.Stderr, "\nTotal: found %s operations for entity: %s\n",
		utils.FormatNumber(entityOps), entityID)

	if entityOps == 0 {
		fmt.Printf("\nNo operations found for this entity!\n")
		return nil
	}

	// Sort timeline
	sort.Slice(result.operations, func(i, j int) bool {
		return result.operations[i].timestamp.Before(result.operations[j].timestamp)
	})

	// Calculate time span
	if len(result.operations) == 0 {
		return nil
	}
	firstOp := result.operations[0].timestamp
	lastOp := result.operations[len(result.operations)-1].timestamp
	timeSpanHours := lastOp.Sub(firstOp).Hours()

	// Analysis and reporting
	fmt.Printf("\n%s\n", strings.Repeat("=", 100))
	fmt.Printf("TIMELINE ANALYSIS FOR: %s\n", entityID)
	fmt.Printf("%s\n", strings.Repeat("=", 100))

	// 1. Summary statistics
	fmt.Printf("\n1. SUMMARY STATISTICS\n")
	fmt.Printf("%s\n", strings.Repeat("-", 100))
	fmt.Printf("Total operations: %s\n", utils.FormatNumber(entityOps))
	fmt.Printf("Time span: %.2f hours (%.2f days)\n", timeSpanHours, timeSpanHours/24.0)

	if timeSpanHours > 0 {
		fmt.Printf("Average rate: %.1f operations/hour (%.2f/minute)\n",
			float64(entityOps)/timeSpanHours,
			float64(entityOps)/timeSpanHours/60.0)
	}
	fmt.Printf("First operation: %s\n", firstOp.Format("2006-01-02 15:04:05"))
	fmt.Printf("Last operation: %s\n", lastOp.Format("2006-01-02 15:04:05"))

	// 2. Operation type distribution
	fmt.Printf("\n2. OPERATION TYPE DISTRIBUTION\n")
	fmt.Printf("%s\n", strings.Repeat("-", 100))
	fmt.Printf("%-30s %-15s %-15s\n", "Operation", "Count", "Percentage")
	fmt.Printf("%s\n", strings.Repeat("-", 100))

	type opEntry struct {
		op    string
		count int
	}
	var sortedOps []opEntry
	for op, count := range result.operationsByType {
		sortedOps = append(sortedOps, opEntry{op, count})
	}
	sort.Slice(sortedOps, func(i, j int) bool {
		return sortedOps[i].count > sortedOps[j].count
	})

	for _, entry := range sortedOps {
		percentage := (float64(entry.count) / float64(entityOps)) * 100.0
		fmt.Printf("%-30s %-15s %-15.2f%%\n", entry.op, utils.FormatNumber(entry.count), percentage)
	}

	// 3. Top paths accessed
	fmt.Printf("\n3. TOP 30 PATHS ACCESSED\n")
	fmt.Printf("%s\n", strings.Repeat("-", 100))
	fmt.Printf("%-70s %-15s %-15s\n", "Path", "Count", "Percentage")
	fmt.Printf("%s\n", strings.Repeat("-", 100))

	type pathEntry struct {
		path  string
		count int
	}
	var sortedPaths []pathEntry
	for path, count := range result.pathsAccessed {
		sortedPaths = append(sortedPaths, pathEntry{path, count})
	}
	sort.Slice(sortedPaths, func(i, j int) bool {
		return sortedPaths[i].count > sortedPaths[j].count
	})

	for i, entry := range sortedPaths {
		if i >= 30 {
			break
		}
		percentage := (float64(entry.count) / float64(entityOps)) * 100.0
		displayPath := entry.path
		if len(displayPath) > 68 {
			displayPath = displayPath[:65] + "..."
		}
		fmt.Printf("%-70s %-15s %-15.2f%%\n", displayPath, utils.FormatNumber(entry.count), percentage)
	}

	// 4. Hourly activity pattern
	fmt.Printf("\n4. HOURLY ACTIVITY PATTERN (Top 30 Hours)\n")
	fmt.Printf("%s\n", strings.Repeat("-", 100))
	fmt.Printf("%-20s %-12s %-10s %-10s %-10s %-10s\n", "Hour", "Total Ops", "read", "update", "list", "Other")
	fmt.Printf("%s\n", strings.Repeat("-", 100))

	type hourEntry struct {
		hour string
		ops  map[string]int
	}
	var sortedHours []hourEntry
	for hour, ops := range result.operationsByHour {
		sortedHours = append(sortedHours, hourEntry{hour, ops})
	}
	sort.Slice(sortedHours, func(i, j int) bool {
		return sortedHours[i].ops["total"] > sortedHours[j].ops["total"]
	})

	for i, entry := range sortedHours {
		if i >= 30 {
			break
		}
		total := entry.ops["total"]
		read := entry.ops["read"]
		update := entry.ops["update"]
		listOps := entry.ops["list"]
		other := total - read - update - listOps

		fmt.Printf("%-20s %-12s %-10s %-10s %-10s %-10s\n",
			entry.hour, utils.FormatNumber(total),
			utils.FormatNumber(read), utils.FormatNumber(update),
			utils.FormatNumber(listOps), utils.FormatNumber(other))
	}

	// 5. Activity distribution by hour of day
	fmt.Printf("\n5. ACTIVITY DISTRIBUTION BY HOUR OF DAY\n")
	fmt.Printf("%s\n", strings.Repeat("-", 100))
	fmt.Printf("%-10s %-15s %-50s\n", "Hour", "Operations", "Bar Chart")
	fmt.Printf("%s\n", strings.Repeat("-", 100))

	maxOpsInHour := 0
	for _, count := range result.hourOfDayStats {
		if count > maxOpsInHour {
			maxOpsInHour = count
		}
	}

	for hour := 0; hour < 24; hour++ {
		ops := result.hourOfDayStats[hour]
		barLength := 0
		if maxOpsInHour > 0 {
			barLength = (ops * 50) / maxOpsInHour
		}
		bar := strings.Repeat("█", barLength)
		fmt.Printf("%02d:00     %-15s %s\n", hour, utils.FormatNumber(ops), bar)
	}

	// 6. Peak activity analysis
	fmt.Printf("\n6. PEAK ACTIVITY WINDOWS\n")
	fmt.Printf("%s\n", strings.Repeat("-", 100))
	fmt.Printf("%-25s %-15s %-20s\n", "5-Minute Window", "Operations", "Rate (ops/sec)")
	fmt.Printf("%s\n", strings.Repeat("-", 100))

	type windowEntry struct {
		window time.Time
		count  int
	}
	var sortedWindows []windowEntry
	for window, count := range result.windowCounts {
		sortedWindows = append(sortedWindows, windowEntry{window, count})
	}
	sort.Slice(sortedWindows, func(i, j int) bool {
		return sortedWindows[i].count > sortedWindows[j].count
	})

	for i, entry := range sortedWindows {
		if i >= 20 {
			break
		}
		rate := float64(entry.count) / 300.0
		fmt.Printf("%-25s %-15s %-20.3f\n",
			entry.window.Format("2006-01-02 15:04"),
			utils.FormatNumber(entry.count), rate)
	}

	// 7. Behavioral patterns
	fmt.Printf("\n7. BEHAVIORAL PATTERNS\n")
	fmt.Printf("%s\n", strings.Repeat("-", 100))

	if timeSpanHours > 1.0 {
		opsPerHour := float64(entityOps) / timeSpanHours
		if opsPerHour > 100.0 {
			fmt.Printf("WARNING: HIGH FREQUENCY: %.0f operations/hour suggests automated polling\n", opsPerHour)
			fmt.Printf("   Recommended action: Implement caching or increase polling interval\n")
		}

		// Check for token lookup abuse
		tokenLookupCount := 0
		for path, count := range result.pathsAccessed {
			if strings.Contains(path, "token/lookup") {
				tokenLookupCount += count
			}
		}

		if tokenLookupCount > 1000 {
			fmt.Printf("WARNING: TOKEN LOOKUP ABUSE: %s token lookups detected\n", utils.FormatNumber(tokenLookupCount))
			fmt.Printf("   Rate: %.1f lookups/hour = %.2f lookups/second\n",
				float64(tokenLookupCount)/timeSpanHours,
				float64(tokenLookupCount)/timeSpanHours/3600.0)
			fmt.Printf("   Recommended action: Implement client-side token TTL tracking\n")
		}

		// Check for path concentration
		if len(sortedPaths) > 0 {
			topPathPct := (float64(sortedPaths[0].count) / float64(entityOps)) * 100.0
			if topPathPct > 30.0 {
				fmt.Printf("WARNING: PATH CONCENTRATION: %.1f%% of operations on single path\n", topPathPct)
				fmt.Printf("   Path: %s\n", sortedPaths[0].path)
				fmt.Printf("   Recommended action: Review why this path is accessed %s times\n", utils.FormatNumber(sortedPaths[0].count))
			}
		}

		// Check for 24/7 activity
		hoursWithActivity := 0
		for h := 0; h < 24; h++ {
			if _, ok := result.hourOfDayStats[h]; ok {
				hoursWithActivity++
			}
		}
		if hoursWithActivity >= 20 {
			fmt.Printf("WARNING: 24/7 ACTIVITY: Active in %d/24 hours\n", hoursWithActivity)
			fmt.Printf("   Suggests automated system or background process\n")
		}
	}

	fmt.Printf("\n%s\n", strings.Repeat("=", 100))

	return nil
}
